### ISSUE TITLE
Dogbone on a component

### ISSUE THREAD

**Issue by rodgah** [DVE2000/Dogbone#9]
Hi,
If I create a new component and then a sketch - > body within that component then try to use your tool it behaves odd. For a very simple shape if I select the body it will perform the dogbone on the first internal corner but then fail (Failed : 3 : object is invalid ..... followed by error in dogbone creation). If I go back and reselect that failed corner by itself and dogbone just the interior edge it works but also errors again. If I initially selected just an interior edge and not the body, same thing .... works but errors. 

Another error I have is creating dogbones on things under 10mm wide. For instance a square with a slot (9mmx9mm) cut into its side will fail with Failed : 5 : Sketch geometry over constrained.

**Comment by coreykovacs** [comment]
I am seeing the same behavior. This is the first time I've tried to use it and i think it's installed correctly. Is there a specific version of python required? I have 3.3.4 installed, F360 v 2.0.1876 free version. Beyond that, thanks for writing this addin, it will be very useful once I can get it working. I have the version pushed this morning with the y-up addition

**Comment by caseycrogers** [comment]
Hello, I just recently updated the code, perhaps I introduced a bug by accident. Can you two send me some screenshots of the geometry and the failures so I can replicate it and find the source of the problem? Alternatively, can you e-mail me the .f3d files of the designs that are failing (caseycrogers@berkeley.edu)?I'm sorry you're having issues with the add-in.

@coreykovacs the versions shouldn't be an issue

**Comment by coreykovacs** [comment]
The geometry for the constraint error was a very simple square in a square, extruded. I tried to apply the dogbones to the internal corners, no joy. I'll send you the archive from my project. I'm not complaining about the problems at all, I know how useful this is when it's working, I just wan to help you debug. I'll send the archive to you direct.

**Comment by Helmi** [comment]
here we go - just copied one out of my current project http://a360.co/1nGw78i

**Comment by caseycrogers** [comment]
Hey guys, I cannot seem to replicate the "Sketch geometry over constrained" issue. After some quick testing, it's pretty clear that the "Failed 3 error" only occurs when applying the add-in to a body that is a part of a non-root component. I'll look into this and see why this is happening, but for now the add-in won't be able to support dog-bones on bodies in components. I'm not sure if this is a result of a Fusion update or because of my most recent push.

@coreykovacs can you send me an .f3d of the design causing the over constrained error?

**Comment by caseycrogers** [comment]
So I have traced the "Failed 3" error to the API's sweep tool support-it does not appear to be working properly on components. I'll post on the Fusion API forum about it and see what the issue is but this seems like a Fusion bug so we might have to wait until the next Fusion update to get support for Dogboning components back.

**Comment by caseycrogers** [comment]
So I was able to write a quick workaround for the problem, however there's no longer any good error handling for when you try to apply a dogbone to impossible geometry and the sweep fails.

All the same, I tested the workaround on both of the designs you guys sent me and it seems to be working. The latest copy is up on github, go ahead and download it and replace your current version and the problem should be fixed. Please still send me the design causing the overconstrained error so I can take a look at that,

**Comment by coreykovacs** [comment]
That's fantastic you see, I'll try it out as soon as I get home. Thank you
so much for all your hard work on this project

**Comment by rodgah** [comment]
wow quick response. Don't get me wrong....this is an amazing tool! I tried the new version and it now works on components, though it shows the same over constrained behaviour in either component or root body. Ive only just started learning fusion so my version is really new. This issue did exist however before your previous update since I originally downloaded it 5 days ago.
Thank you for your help.

[dogbone fail.zip](https://github.com/caseycrogers/Dogbone/files/111525/dogbone.fail.zip)

**Comment by coreykovacs** [comment]
I agree, the turn around is amazing, thanks.  I got home and tested the update. Interestingly, when I picked an entire panel, it worked, sort of. What ended up getting the dogbones, was the "bottom". Turns out for some reason my project is "y-up" which is something I thought I changed before creating this thing. Now it works as expected, with one added "feature", which could be caused by something dumb I did. When it does it's thing, the parts are all moved around. They are correct, just moved. Weird.

**Comment by coreykovacs** [comment]
More...  When it moves things around (I have multiple panels ) what ended up happening was the overlap caused holes to be placed onto parts in locations that did not need them. I'll send you a couple of screen shots.

**Comment by coreykovacs** [comment]
Casey, if you look below, you'll see what i was trying to explain.

The first picture, is the original layout.
Second picture is after the first attempt. it scrambled everything up and
the panel I selected by luck ended up by itself so no corruption of the
other panels. The third shot, is what happened when I selected a panl which
was mixed in. Come to think about, this is probably predictable. So I
"moved" and then dogboned the large panel, it moved to the "cluster" again
and it too resulted in other panels having holes "drilled into them.

I am totally down with the idea that my model simply sucks and I did
something wrong. :)

I'll try to create something with the axis oriented sanely and try again.


[image: Inline image 1][image: Inline image 2][image: Inline image 3]

**Comment by coreykovacs** [comment]
More...  When it moves things around (I have multiple panels ) what ended up happening was the overlap caused holes to be placed onto parts in locations that did not need them. I'll send you a couple of screen shots.
![before](https://cloud.githubusercontent.com/assets/16987784/12706037/0c1dab0e-c831-11e5-82ad-c9399223ce6a.PNG)

![after-1](https://cloud.githubusercontent.com/assets/16987784/12706038/14efdb6c-c831-11e5-8f2d-ad15547d145d.PNG)

![after-2](https://cloud.githubusercontent.com/assets/16987784/12706040/1923b44c-c831-11e5-98e1-57b1e97ec582.PNG)

**Comment by caseycrogers** [comment]
Hey Corey, in the far right of the fusion tool bar click "take snapshot". That should keep the panels from moving.

**Comment by coreykovacs** [comment]
Thanks will do!

Corey

**Comment by coreykovacs** [comment]
That did the trick, thanks!

**Comment by Helmi** [comment]
sorry, did find the time for testing today, thanks for your great and quick response anyhow. Will test as soon as i get to it.

**Comment by rodgah** [comment]
Hi Helmi can you try the file I attached above to see if you get the same over constrained error please.

**Comment by caseycrogers** [comment]
Hey @rodgah I'll download your design later today and see if I can figure out what's going wrong.

**Comment by rodgah** [comment]
Hey Casey, thanks in your own time no rush. I was just interested to see if my issue wasn't isolated to just me.

**Comment by caseycrogers** [comment]
@rodgah So I downloaded your design and did get the failure. The conditions under which it fails seem pretty arbitrary and I can't really figure out what's going on, so I posted on the Fusion 360 forum about it, I'll keep you updated.

**Comment by Helmi** [comment]
@caseycrogers looks like it's working - great, thanks.

Is there any way to make the addin more "talky"? At first i thought it's not working until i noticed it's still running in the background and it needs its time. I don't know about the possibilities with the API but if there's any chance to keep the window open where you enter your values and indicate the script is still running that would be a nice addition.

**Comment by caseycrogers** [comment]
In my most recent update I made it display each dogbone as it's created (before it finished all the dogbones and then displayed the results). This hopefully makes it a little bit more clear that the add-in is working.

Additionally, in the bottom right corner, there's a green bar that shows the progress on each individual task the add-in is completing. If you see the green bar flashing and moving you know the add-in is running.

Other than that, unfortunately Fusion 360's API doesn't really give me other options for communicating progress to the user.

**Comment by caseycrogers** [comment]
@rodgah So I was able to narrow down the issue to a small bug in the Fusion symmetry constraint API and used a quick work around. The problem should now be fixed, go ahead and download the latest version of dogbone.py and tell me if you still have any trouble!

**Comment by rodgah** [comment]
@caseycrogers I tried re downloading but it didn't change.....then I noticed the file hasn't been updated did you commit?

**Comment by PHermans1** [comment]
How to use this add-in?
I am new to Fushion 360 and downloaded this add-in, placed it in the appropriate folder. When I restarted it appears in add-in dialog and on menu bar. I tried using it, and get a dialog with several options that are not explained anywhere (that I can see) and when I pick my component and click OK, nothing appears to happen.  Not sure why not etc, or if there is any where I can look to see what is happening, but so far finding it to be a bit frustrating.

Does it work on components?
